static/js/import_wizard/utils/detection.ts

```diff
@@ -0,0 +1,216 @@
+/**
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import _ from "lodash";
+
+import {
+  Column,
+  MappedThing,
+  Mapping,
+  MappingType,
+  MappingVal,
+  TypeProperty,
+} from "../types";
+
+const COLUMN_API_KEY = "column";
+const COLUMN_ID_API_KEY = "id";
+const COLUMN_HEADER_API_KEY = "header";
+const COLUMN_IDX_API_KEY = "column_idx";
+const DCID_API_KEY = "dcid";
+const DISP_NAME_API_KEY = "display_name";
+const PLACETYPE_API_KEY = "place_type";
+const PLACEPROP_API_KEY = "place_property";
+const DCTYPE_API_KEY = "dc_type";
+const DCPROP_API_KEY = "dc_property";
+const HEADER_API_KEY = "headers";
+
+function isDCTypeOrProperty(entity): boolean {
+  if (_.has(entity, DCID_API_KEY) && _.has(entity, DISP_NAME_API_KEY)) {
+    return true;
+  }
+  return false;
+}
+
+function isColumn(candidate): boolean {
+  if (
+    _.has(candidate, COLUMN_ID_API_KEY) &&
+    _.has(candidate, COLUMN_HEADER_API_KEY) &&
+    _.has(candidate, COLUMN_IDX_API_KEY)
+  ) {
+    return true;
+  }
+  return false;
+}
+
+/**
+ * Given the API response from the detection server API,
+ * parse it in to a Mapping structure.
+ *
+ * @param detectedResponse is the Object response from the server API.
+ *
+ * @returns a Mapping structure which has details of all the detected
+ *  columns. If there are any unexpected validation errors, that particular
+ *  MappingVal is skipped.
+ */
+export function parseDetectionAPIResponse(detectedResponse): Mapping {
+  const mToReturn: Mapping = new Map<MappedThing, MappingVal>();
+
+  // Some validations.
+  for (const [key, mVal] of Object.entries(detectedResponse)) {
+    // Keys must be among the MappedThing enum.
+    if (!_.includes(Object.values(MappedThing), key)) {
+      continue;
+    }
+    // Key "type" must be present and the value must be among MappingType enum.
+    if (
+      !_.includes(Object.keys(mVal), "type") ||
+      !_.includes(Object.values(MappingType), mVal["type"])
+    ) {
+      continue;
+    }
+    const mValValidated: MappingVal = { type: mVal["type"] };
+
+    // If key "column" is present then the value must be of type Column.
+    let col: Column;
+    if (_.includes(Object.keys(mVal), COLUMN_API_KEY) && mVal[COLUMN_API_KEY]) {
```

Comment:
rather than doing these checks on the client based on response keys, can the API return an enum with mapping type? that would simplify the code here a lot. we don't need this API to be very general purpose and can tailor it to our use case.

separately though, could this simply be:
`_.has(mVal, COLUMN_API_KEY)`
to avoid using `Object.keys` repeatedly

Reply:
Made the suggested change with eliminating Object.keys(). However, I think I'll have to punt the other suggestion to a later PR after discussion--it's a fairly big change and will need more care and thought. Let me know if you have strong feelings against?

Reply:
Sorry I missed your question. We can do it as a followup

Reply:
Thanks.